File: crypto_auto_trade/validation.py
from __future__ import annotations

from crypto_auto_trade.backtest import BacktestConfig, Backtester, forward_test
from crypto_auto_trade.models import Candle
from crypto_auto_trade.strategies import build_strategy, strategy_names


def run_validation_matrix(candles: list[Candle], iterations: int = 200, trailing_stop_pct: float = 0.05) -> dict[str, object]:
    if len(candles) < 120:
        raise ValueError("validation needs at least 120 candles")
    names = strategy_names()
    rows: list[dict[str, object]] = []
    config = BacktestConfig(trailing_stop_pct=trailing_stop_pct)
    for i in range(iterations):
        name = names[i % len(names)]
        window = min(len(candles), 120 + (i * 17) % max(1, len(candles) - 119))
        start = (i * 13) % max(1, len(candles) - window + 1)
        result = Backtester(build_strategy(name), config).run(candles[start : start + window])
        verdict = "healthy" if result.total_return > 0 and result.max_drawdown < 0.25 else "watch"
        if result.max_drawdown >= 0.35:
            verdict = "risk_high"
        rows.append(
            {
                "strategy": name,
                "window": window,
                "start": start,
                "total_return": round(result.total_return, 6),
                "max_drawdown": round(result.max_drawdown, 6),
                "sharpe_like": round(result.sharpe_like, 6),
                "trade_count": len(result.trades),
                "trailing_stop_count": result.as_dict()["trailing_stop_count"],
                "verdict": verdict,
            }
        )
    summary = summarize(rows)
    return {"iterations": iterations, "trailing_stop_pct": trailing_stop_pct, "summary": summary, "best": summary[0] if summary else None, "rows": rows}


def summarize(rows: list[dict[str, object]]) -> list[dict[str, object]]:
    out: list[dict[str, object]] = []
    for name in sorted({str(r["strategy"]) for r in rows}):
        selected = [r for r in rows if r["strategy"] == name]
        positives = sum(1 for r in selected if float(r["total_return"]) > 0)
        healthy = sum(1 for r in selected if r["verdict"] == "healthy")
        avg_return = sum(float(r["total_return"]) for r in selected) / len(selected)
        avg_drawdown = sum(float(r["max_drawdown"]) for r in selected) / len(selected)
        avg_sharpe = sum(float(r["sharpe_like"]) for r in selected) / len(selected)
        score = avg_return * 100 + avg_sharpe - avg_drawdown * 50 + healthy / len(selected) * 10
        out.append(
            {
                "strategy": name,
                "runs": len(selected),
                "positive_rate": round(positives / len(selected), 4),
                "healthy_rate": round(healthy / len(selected), 4),
                "avg_return": round(avg_return, 6),
                "avg_drawdown": round(avg_drawdown, 6),
                "avg_sharpe_like": round(avg_sharpe, 6),
                "total_trailing_stops": sum(int(r["trailing_stop_count"]) for r in selected),
                "selection_score": round(score, 6),
            }
        )
    return sorted(out, key=lambda r: float(r["selection_score"]), reverse=True)


def compare_all_strategies(candles: list[Candle], trailing_stop_pct: float = 0.05) -> list[dict[str, object]]:
    config = BacktestConfig(trailing_stop_pct=trailing_stop_pct)
    rows = [Backtester(build_strategy(name), config).run(candles).as_dict() for name in strategy_names()]
    return sorted(rows, key=lambda r: (float(r["total_return"]), -float(r["max_drawdown"])), reverse=True)


def forward_all_strategies(candles: list[Candle], trailing_stop_pct: float = 0.05) -> list[dict[str, object]]:
    config = BacktestConfig(trailing_stop_pct=trailing_stop_pct)
    return [forward_test(build_strategy(name), candles, config) for name in strategy_names()]


def select_best_strategy(candles: list[Candle], iterations: int = 300, trailing_stop_pct: float = 0.05) -> dict[str, object]:
    matrix = run_validation_matrix(candles, iterations=iterations, trailing_stop_pct=trailing_stop_pct)
    best = matrix.get("best")
    return {
        "method": "rolling-window validation across all registered strategy variants",
        "note": "This is not a profit guarantee. It selects the strongest historical candidate under the current sample/data and trailing stop setting.",
        "best": best,
        "strategy_count": len(strategy_names()),
        "iterations": iterations,
        "trailing_stop_pct": trailing_stop_pct,
    }

File: worker/src/validation.ts
import { Candle } from "./indicators";
import { strategyNames } from "./strategies";
import { BacktestConfig, defaultConfig, runBacktest, forwardTest } from "./backtest";

const r = (v: number, n: number): number => {
  const f = 10 ** n;
  return Math.round(v * f) / f;
};

export interface MatrixRow {
  strategy: string;
  window: number;
  start: number;
  total_return: number;
  max_drawdown: number;
  sharpe_like: number;
  trade_count: number;
  trailing_stop_count: number;
  verdict: string;
}

export interface SummaryRow {
  strategy: string;
  runs: number;
  positive_rate: number;
  healthy_rate: number;
  avg_return: number;
  avg_drawdown: number;
  avg_sharpe_like: number;
  total_trailing_stops: number;
  selection_score: number;
}

export function runValidationMatrix(candles: Candle[], iterations = 200, trailingStopPct = 0.05): Record<string, unknown> {
  if (candles.length < 120) throw new Error("validation needs at least 120 candles");
  const names = strategyNames();
  const rows: MatrixRow[] = [];
  const config = defaultConfig(trailingStopPct);
  for (let i = 0; i < iterations; i++) {
    const name = names[i % names.length];
    const window = Math.min(candles.length, 120 + ((i * 17) % Math.max(1, candles.length - 119)));
    const start = (i * 13) % Math.max(1, candles.length - window + 1);
    const result = runBacktest(name, candles.slice(start, start + window), config);
    let verdict = result.total_return > 0 && result.max_drawdown < 0.25 ? "healthy" : "watch";
    if (result.max_drawdown >= 0.35) verdict = "risk_high";
    rows.push({
      strategy: name,
      window,
      start,
      total_return: r(result.total_return, 6),
      max_drawdown: r(result.max_drawdown, 6),
      sharpe_like: r(result.sharpe_like, 6),
      trade_count: result.trade_count,
      trailing_stop_count: result.trailing_stop_count,
      verdict,
    });
  }
  const summary = summarize(rows);
  return { iterations, trailing_stop_pct: trailingStopPct, summary, best: summary[0] ?? null, rows };
}

export function summarize(rows: MatrixRow[]): SummaryRow[] {
  const out: SummaryRow[] = [];
  const names = [...new Set(rows.map((r) => r.strategy))].sort();
  for (const name of names) {
    const selected = rows.filter((r) => r.strategy === name);
    const positives = selected.filter((r) => r.total_return > 0).length;
    const healthy = selected.filter((r) => r.verdict === "healthy").length;
    const avgReturn = selected.reduce((a, b) => a + b.total_return, 0) / selected.length;
    const avgDrawdown = selected.reduce((a, b) => a + b.max_drawdown, 0) / selected.length;
    const avgSharpe = selected.reduce((a, b) => a + b.sharpe_like, 0) / selected.length;
    const score = avgReturn * 100 + avgSharpe - avgDrawdown * 50 + (healthy / selected.length) * 10;
    out.push({
      strategy: name,
      runs: selected.length,
      positive_rate: r(positives / selected.length, 4),
      healthy_rate: r(healthy / selected.length, 4),
      avg_return: r(avgReturn, 6),
      avg_drawdown: r(avgDrawdown, 6),
      avg_sharpe_like: r(avgSharpe, 6),
      total_trailing_stops: selected.reduce((a, b) => a + b.trailing_stop_count, 0),
      selection_score: r(score, 6),
    });
  }
  return out.sort((a, b) => b.selection_score - a.selection_score);
}

export function compareAllStrategies(candles: Candle[], trailingStopPct = 0.05): Record<string, unknown>[] {
  const config = defaultConfig(trailingStopPct);
  const rows = strategyNames().map((name) => runBacktest(name, candles, config) as unknown as Record<string, unknown>);
  return rows.sort((a, b) => {
    const ar = a.total_return as number, br = b.total_return as number;
    if (br !== ar) return br - ar;
    return (a.max_drawdown as number) - (b.max_drawdown as number);
  });
}

export function forwardAllStrategies(candles: Candle[], trailingStopPct = 0.05): Record<string, unknown>[] {
  const config = defaultConfig(trailingStopPct);
  return strategyNames().map((name) => forwardTest(name, candles, config));
}

export function selectBestStrategy(candles: Candle[], iterations = 300, trailingStopPct = 0.05): Record<string, unknown> {
  const matrix = runValidationMatrix(candles, iterations, trailingStopPct);
  return {
    method: "rolling-window validation across all registered strategy variants",
    note: "This is not a profit guarantee. It selects the strongest historical candidate under the current sample/data and trailing stop setting.",
    best: (matrix as { best: unknown }).best,
    strategy_count: strategyNames().length,
    iterations,
    trailing_stop_pct: trailingStopPct,
  };
}

export { BacktestConfig };
